lib/ClangImporter/ImportMacro.cpp

```diff
@@ -455,17 +521,46 @@ static ValueDecl *importMacro(ClangImporter::Implementation &impl,
     clang::Token const &first = tokenI[0];
     clang::Token const &second = tokenI[1];
 
-    if (isSignToken(first) && second.is(clang::tok::numeric_constant))
-      return importNumericLiteral(impl, DC, macro, name, &first, second, ClangN,
-                                  castType);
+    if (isSignToken(first) && second.is(clang::tok::numeric_constant)) {
+      ValueDecl *importedNumericLiteral = importNumericLiteral(
+          impl, DC, macro, name, &first, second, ClangN, castType);
+      if (!importedNumericLiteral) {
+        emitMacroDiagnostic(impl,
+                            Diagnostic(diag::macro_not_imported, name.str()),
+                            macro->getDefinitionLoc(), ClangN);
+        emitMacroDiagnostic(
+            impl, Diagnostic(diag::macro_not_imported_invalid_numeric_literal),
+            second.getLocation(), ClangN);
+      }
+      return importedNumericLiteral;
+    }
 
     // We also allow @"string".
-    if (first.is(clang::tok::at) && isStringToken(second))
-      return importStringLiteral(impl, DC, macro, name, second,
-                                 MappedStringLiteralKind::NSString, ClangN);
+    if (first.is(clang::tok::at) && isStringToken(second)) {
+      ValueDecl *importedStringLiteral =
+          importStringLiteral(impl, DC, macro, name, second,
+                              MappedStringLiteralKind::NSString, ClangN);
+      if (!importedStringLiteral) {
+        emitMacroDiagnostic(impl,
+                            Diagnostic(diag::macro_not_imported, name.str()),
+                            macro->getDefinitionLoc(), ClangN);
+        emitMacroDiagnostic(
+            impl, Diagnostic(diag::macro_not_imported_invalid_string_literal),
+            second.getLocation(), ClangN);
+      }
+      return importedStringLiteral;
+    }
+
     break;
   }
   case 3: {
+    // Regect 3-token macros invoking nullary functions like the following:
```

Comment:
Typo: `regect` → `reject`

Also, this is new behaviour, right? I think I understand why you do it, but because the ad hoc way the macros are parsed I am not sure if rejecting up here is the right way.

Reply:
It is new behavior, would it make sense to you to change the below comment to produce a `macro_not_imported_unsupported_structure` and let it fall through instead? The outcome for this type of macro is equivalent.